admittance_controller/include/admittance_controller/admittance_rule_impl.hpp

```diff
@@ -0,0 +1,534 @@
+    state_message_.admittance_rule_calculated_values.positions.resize(6, 0.0);
+    state_message_.admittance_rule_calculated_values.velocities.resize(6, 0.0);
+    state_message_.admittance_rule_calculated_values.accelerations.resize(6, 0.0);
+
+
+    // admittance state vectors in joint space
+    joint_pos_ = Eigen::Matrix<double, Eigen::Dynamic, 1, Eigen::ColMajor>(num_joints_);
+    joint_vel_ = Eigen::Matrix<double, Eigen::Dynamic, 1, Eigen::ColMajor>(num_joints_);
+    joint_acc_ = Eigen::Matrix<double, Eigen::Dynamic, 1, Eigen::ColMajor>(num_joints_);
+
+    // transforms
+    ee_transform_.setIdentity();
+    reference_ft_transform_.setIdentity();
+    sensor_transform_.setIdentity();
+    control_transform_.setIdentity();
+    cog_transform_.setIdentity();
+    world_transform_.setIdentity();
+    ft_transform_.setIdentity();
+
+    // admittance values
+    admittance_position_.setIdentity();
+    admittance_velocity_.setZero();
+    admittance_acceleration_.setZero();
+    wrench_world_.setZero();
+    jacobian_ = Eigen::Matrix<double, 6, Eigen::Dynamic, Eigen::ColMajor>(6, num_joints_);
+    jacobian_.setZero();
+    identity = Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::ColMajor>(num_joints_, num_joints_);
+    identity.setIdentity();
+    damping_.setZero();
+    mass_.setOnes();
+    mass_inv_.setZero();
+    stiffness_.setZero();
+    selected_axes_.setZero();
+
+    // parameters
+    if (parameter_handler_->params_.enable_parameter_update_without_reactivation_) {
+      // point to the dynamic data
+      parameters_ = &parameter_handler_->params_;
+    } else {
+      // make a copy of the data
+      parameters_copy = parameter_handler_->params_;
+      parameters_ = &parameters_copy;
+    }
+
+    return controller_interface::return_type::OK;
+  }
+
+  // Update from reference joint states
+  controller_interface::return_type AdmittanceRule::update(
+      const trajectory_msgs::msg::JointTrajectoryPoint &current_joint_state,
+      const geometry_msgs::msg::Wrench &measured_wrench,
+      const trajectory_msgs::msg::JointTrajectoryPoint &reference_joint_state,
+      const rclcpp::Duration &period,
+      trajectory_msgs::msg::JointTrajectoryPoint &desired_joint_state) {
+
+    double dt = period.seconds() + ((double) period.nanoseconds()) * 1E-9;
+
+    // update param values
+    ee_weight.setZero();
+    ee_weight[2] = -parameters_->gravity_compensation_.CoG_.force_;
+    vec_to_eigen(parameters_->gravity_compensation_.CoG_.pos_, cog_);
+    vec_to_eigen(parameters_->admittance_.mass_, mass_);
+    vec_to_eigen(parameters_->admittance_.stiffness_, stiffness_);
+    vec_to_eigen(parameters_->admittance_.selected_axes_, selected_axes_);
+
+    for (auto i = 0ul; i < 6; ++i) {
+      mass_inv_[i] = 1.0 / parameters_->admittance_.mass_[i];
+      damping_[i] = parameters_->admittance_.damping_ratio_[i] * 2 * sqrt(mass_[i] * stiffness_[i]);
+    }
```

Comment:
Maybe we don't want the parameters to be dynamic, but that is a design decision. I know users will want to enable and disable admittance for certain parts of a trajectory. Its true, they can restart the admittance controller, but they will be forced including waiting way points during their tasks. That seems very unnecessary. Perhaps we should only allow certain parameters to be dynamic?

Reply:
TBH, I can not imagine scenario when you want to have admittance only on the part of the trajectory. And trajectory means a continuous movement which is not allowed to be interrupted. Switching controllers between trajectory segments shouldn't be any isse.  And that you want to control that using parameters that are everything but not real-time safe or anywhere near determinism in communication. So this will in any case not work with this approach.

We already talked about that, to have such one would need to extend Trajectory message with force information.